🪐 Per-Output AST Representation for Code Cell Outputs

rowanc1 opened the pull request:
Extend the MyST AST to represent individual code cell outputs as separate AST nodes, enabling output content to participate in MyST builds (e.g., generate and consume reference labels).

[![Preview on MyST Sandbox](https://img.shields.io/badge/Preview%20on%20MyST%20Sandbox-gray?style=for-the-badge&link=https://mystmd.org/sandbox?url=https://github.com/jupyter-book/myst-enhancement-proposals/blob/mep/outputs/meps/mep-0004.md?raw=true)](https://mystmd.org/sandbox?url=https://github.com/jupyter-book/myst-enhancement-proposals/blob/mep/outputs/meps/mep-0004.md?raw=true)

## Authors

- @agoose77
- @rowanc1

## Issue

- https://github.com/jupyter-book/mystmd/issues/1026

choldgraf commented on meps/mep-0004.md:
```diff
@@ -0,0 +1,273 @@
+
+### Use Cases
+
+This enhancement enables several important use cases:
```
```suggestion
This enhancement lays the foundation to enable several important use cases (it does not solve for those use-cases now, but is a required step):
```

choldgraf replied:
I think this is true - we still need to do the follow up work for the other use-cases, right? This just makes that clear

choldgraf commented on meps/mep-0004.md:
````diff
@@ -0,0 +1,273 @@
+**New Structure (Version 3):**
+
+```typescript
+type Output = {
````
```suggestion
// Outputs (below) contains one or more Output nodes
type Output = {
```

choldgraf replied:
it confused me to see both of them at first, I thought you meant it's called 'Output' when there's one node, and 'Outputs' when there are more than one. But I think instead you're saying they're nested so I just tried to clarify

choldgraf reviewed (approved):
This is great - I had a few questions to clarify but IMO it could be merged as-is whenever you two are happy w/ the language.

Honestly my only critique is that this MEP is _very_ thorough - I worry it's almost too high of a bar to set for other MEPs but that's not really a problem per-se, just a thought I was having as I was reading.

agoose77 commented on meps/mep-0004.md:
````diff
@@ -0,0 +1,273 @@
+}
+```
+
+## Examples
````
I think we should produce multiple MEPs. In this one, it's just AST reworking. 

We should remove this - it's not part of *this* MEP. Well make another MEP to handle outputs parsing.

stevejpurves commented on meps/mep-0004.md:
```diff
@@ -0,0 +1,188 @@
+
+### Theme Considerations
+
+Themes and renderers will need to be updated to handle the new `Outputs` container node and iterate over `Output` children. However, the rendering output should remain consistent, as the logical grouping of outputs is preserved. Currently there is coupling between the HTML site (the result of `myst start`) and the CLI. Version mismatches can result in outputs not being shown in older themes; we a ...
```
I feel like this is missing something on:

* the differences in impact on i. the local dev/preview experience for authors, ii. building of static sites and iii. dynamic theme users and deployed-vs-new content -- these are distinct use cases that mystmd supports

* the impact of these changes are worse for themes that have not yet adopted the `myst-migrate` infrastrucure -- which at the time of this change are all themes. Themes that had adopted `myst-migrate` will fare better

choldgraf replied:
> the impact of these changes are worse for themes that have not yet adopted the myst-migrate infrastrucure -- which at the time of this change are all themes.

this made me laugh 😅

agoose77 commented on meps/mep-0004.md:
```diff
@@ -0,0 +1,195 @@
+
+### Backward Compatibility
+
+:::{warning}
+Upgrading from V2 to V3 is a lossy process if the number of outputs is greater than one. It is not possible to exactly determine from a flat list of outputs which AST node correlates to which IOutput object. As such, we choose to drop them.
+:::
```
@rowanc1 we've spoken about this before, but I want to flag it again.

I don't think it affects many documents, but you should be mindful of this when upgrading your own ASTs. If you have e.g. a code-cell generating N figures, you'll need to ensure that the upgrade is not lossy. 

In simple english:
- All placeholders are preserved, as children of `Outputs`
- Non-placeholders are only preserved if they are _the only child_ of `Output`.

Explicitly:
```yaml
type: output
data: 
  - foo
children:
  - type: image
    placeholder: true
```
is fine, as is 
```yaml
type: output
data: 
  - foo
children:
  - type: image
```
but
```yaml
type: output
data: 
  - foo
children:
  - type: image
    placeholder: true
  - type: image
```
and
```yaml
type: output
data: 
  - foo
  - bar
children:
  - type: image
  - type: image
```
will involve loss -- we can't assume that each child maps to each IOutput, so we don't and drop them.

choldgraf commented on meps/mep-0004.md:
```diff
@@ -0,0 +1,211 @@
+authors:
+  - agoose77
+  - rowanc1
+tags:
+  - Active
```
```suggestion
```

We hit some confusion about whether the frontmatter is the source of truth, or if the github label is the source of truth. @rowanc1 and I suggested removing the tag metadata and using github PR labels for now, and when the PR is resolved, that's the point where we encode state into the frontmatter of the document.

That way, the state management is more fluid and w/ a better UX while it's changing, and when it's stabilized (ie, the MEP is completed) then we can put it in a more stable place

rowanc1 commented:
Hi all - we have just set the Per-Jupyter Output AST representation to `Active`.

We are planning to leave this open for about Jan 15 if folks have comments. Much of this has been implemented in mystmd and is opening up exciting new capabilities for per-output parsing etc. with myst-native AST (e.g. `display(md)` from Jupyter nodes).

Looking forward to feedback and continuing to refine our MEP process and make it useful for our team!

agoose77 committed:
Set status to active

agoose77 committed:
Change use case (3)

I don't think text-based AST generation is that useful. It's trivial to
tell people to use the display protocol for this, which permits Python
etc. to output structured data (JSON) that is validated.

rowanc1 committed:
Update meps/mep-0004.md

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

agoose77 committed:
wip: tweaks

rowanc1 committed:
Make shorter and more focused

agoose77 committed:
Add note about lossy upgrade

rowanc1 committed:
🪐 Per-Output AST Representation for Code Cell Outputs

agoose77 committed:
Puncutation and detail

agoose77 committed:
Update meps/mep-0004.md

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

agoose77 committed:
Add abbreviations

agoose77 committed:
Add detail on upgrades

rowanc1 commented on meps/mep-0004.md:
```diff
@@ -0,0 +1,211 @@
+- Cross-reference resolution
+- Label generation
+- Citation processing
+- Other MyST transforms
```
@stefanv — remove this section as it is out of scope?

```suggestion
```

rowanc1 committed:
Remove out of scope content
